src/clj/puppetlabs/puppetserver/certificate_authority.clj

```diff
@@ -747,9 +766,33 @@
   should be copied to."
   ([cacrl :- schema/Str
     localcacrl :- schema/Str]
-    (when (fs/exists? cacrl)
-      (ks/mkdirs! (fs/parent localcacrl))
-      (fs/copy cacrl localcacrl))))
+   (when (not= cacrl localcacrl)
+     (let [max-attempts 5]
+       (when (and (fs/exists? cacrl))
+         (ks/mkdirs! (fs/parent localcacrl))
+         ;; Make up to 'max-attempts' tries to copy the cacrl file to the
+         ;; localcacrl file.  The content of the cacrl file is read into memory
+         ;; and parsed for valid CRL pem content during each attempt.  The content
+         ;; in memory is written to the localcacrl file only if it is valid.  This
+         ;; validation is done to protect against a partially written (which could
+         ;; happen if an asynchronous revocation is in process) or corrupt
+         ;; cacrl file being copied.
+         (loop [attempts-left max-attempts
+                ca-crl-last-modified (fs/mod-time cacrl)]
+           (let [cacrl-as-string (slurp cacrl)]
+             (when-let [write-exception (write-local-cacrl!
+                                         localcacrl
+                                         cacrl-as-string)]
+               (if (zero? attempts-left)
+                 (log/error (i18n/trs "Unable to synchronize crl file {0} to {1}: {2}"
+                                      cacrl localcacrl (.getMessage write-exception)))
+                 (do
+                   (Thread/sleep 1000)
+                   (let [new-ca-crl-last-modified (fs/mod-time cacrl)]
+                     (recur (if (= ca-crl-last-modified new-ca-crl-last-modified)
```

Comment:
@camlow325 was there a specific concern you were guarding against here with the last modified check? It seems like even if we log an error here and move along, that when the modification finishes it will just trigger a new fs-watcher event and try again.
I'm inclined to remove the modification time check and simplify the logic here.